README.md
# 规范指北

1. 用命令行工具打开你的项目目录
2. 使用以下命令安装必要的包：
    ```bash
    npm install --save-dev @commitlint/cli @commitlint/config-conventional conventional-changelog commitizen cz-git husky
    ```
3. 配置 `commitlint` 规则。

   在项目根目录中创建一个名为 `.commitlintrc.js` 的文件，并添加以下内容：

   ```js
   // .commitlintrc.js
   const {execSync} = require("child_process");
   const util = require("util");
   
   // 获取用户名和当前日期
   const author = execSync('git config user.name').toString().replace(/(\r\n\t|\n|\r\t)/g, '').trim()
   const date = new Date().toLocaleDateString()
   
   /** @type {import('cz-git').UserConfig} */
   
   const issuePrefixes = [
     // 如果使用 tapd 作为开发管理
     {value: 'bug', name: 'bug:     标记缺陷已完成'},
     {value: 'story', name: 'story:   标记需求已完成'}
   ];
   
   module.exports = {
     extends: ['@commitlint/config-conventional'],
     rules: {
       'subject-min-length': [2, 'always', 2],
       'subject-empty': [2, 'never'],
     },
     prompt: {
       alias: {fd: 'docs: fix typos'},
       messages: {
         type: '选择你要提交的类型 :',
         scope: '选择一个提交范围（可选）:',
         customScope: '请输入自定义的提交范围 :',
         subject: '填写简短精炼的变更描述 :\n',
         body: '填写更加详细的变更描述（可选）。使用 "|" 换行 :\n',
         breaking: '列举非兼容性重大的变更（可选）。使用 "|" 换行 :\n',
         footerPrefixesSelect: '选择关联tapd内容（可选）:',
         customFooterPrefix: '输入自定义tapd前缀 :',
         footer: '输入对应id (可选) :\n',
         confirmCommit: '是否提交或修改commit ?',
         generatingByAI: '正在用AI生成你的提交信息...',
         generatedSelectByAI: '选择你认为合适的:',
       },
       types: [
         {value: 'feat', name: 'feat:     新增功能 | A new feature'},
         {value: 'fix', name: 'fix:      修复缺陷 | A bug fix'},
         {value: 'docs', name: 'docs:     文档更新 | Documentation only changes'},
         {value: 'style', name: 'style:    代码格式 | Changes that do not affect the meaning of the code'},
         {value: 'refactor', name: 'refactor: 代码重构 | A code change that neither fixes a bug nor adds a feature'},
         {value: 'perf', name: 'perf:     性能提升 | A code change that improves performance'},
         {value: 'test', name: 'test:     测试相关 | Adding missing tests or correcting existing tests'},
         {value: 'build', name: 'build:    构建相关 | Changes that affect the build system or external dependencies'},
         {value: 'ci', name: 'ci:       持续集成 | Changes to our CI configuration files and scripts'},
         {value: 'revert', name: 'revert:   回退代码 | Revert to a commit'},
         {value: 'chore', name: 'chore:    其他修改 | Other changes that do not modify src or test files'},
       ],
       useEmoji: false,
       emojiAlign: 'center',
       themeColorCode: '38;5;043',
       scopes: ['framework', 'lobby', 'music', 'health', 'work'],
       allowCustomScopes: true,
       allowEmptyScopes: true,
       customScopesAlign: 'bottom',
       customScopesAlias: 'custom',
       emptyScopesAlias: 'empty',
       upperCaseSubject: false,
       markBreakingChangeMode: false,
       allowBreakingChanges: ['feat', 'fix'],
       breaklineNumber: 100,
       breaklineChar: '|',
       skipQuestions: [],
       issuePrefixes: issuePrefixes,
       customIssuePrefixAlign: 'bottom',
       emptyIssuePrefixAlias: 'skip',
       customIssuePrefixAlias: 'custom',
       allowCustomIssuePrefix: false,
       allowEmptyIssuePrefix: true,
       confirmColorize: true,
       maxHeaderLength: Infinity,
       maxSubjectLength: Infinity,
       minSubjectLength: 0,
       scopeOverrides: undefined,
       defaultBody: '',
       defaultIssues: '',
       defaultScope: '',
       defaultSubject: '',
       aiNumber: 5,
       aiQuestionCB: ({
                        maxSubjectLength,
                        diff
                      }) => `用完整句子为以下 Git diff 代码写一个有见解并简洁的 Git 中文提交消息，不加任何前缀，并且内容不能超过 ${maxSubjectLength} 个字符: \`\`\`diff\n${diff}\n\`\`\``,
       formatMessageCB: (messageMod => {
   
         // 生成类型和范围
         let scope = (messageMod.scope === undefined || messageMod.scope.length === 0) ? '' : util.format('(%s)', messageMod.scope)
   
         let typeScope = util.format('%s%s:', messageMod.type, scope)
   
         // 生成header
         let subjects = []
         subjects.push(author, date, messageMod.subject)
         subjects = subjects.filter(Boolean)
         const subject = subjects.join('  ')
   
         // 生成header
         const header = util.format('%s %s', typeScope, subject)
   
         // 生成 body 和 breaking
         const body = messageMod.body
         const breaking = (messageMod.breaking === undefined || messageMod.breaking.length === 0) ? undefined : '重大更新：' + messageMod.breaking
   
         // 生成 footer
         let footer
         if (messageMod.footer !== undefined && messageMod.footer.length > 0) {
           try {
             const valuesArray = issuePrefixes.map(obj => obj.value);
             const values = valuesArray.join('|');
   
             const issuePattern = new RegExp(`(?:${values})`, 'g');
             const issue = messageMod.footer.match(issuePattern)
   
             // 尝试替换掉 颜色符号
             let idString = messageMod.footer.replaceAll(/\x1B\[\w*m/g, '')
             // 替换回车
             idString = idString.replaceAll('\n', '')
   
             const idPattern = /\b\d+\b/g;
             const ids = idString.match(idPattern);
             footer = util.format('--%s=%s --user=%s', issue[0], ids.join(','), author)
           } catch (ex) {
             console.log('生成 foot 失败：' + ex)
   
             footer = messageMod.footer.replaceAll('\n', '')
           }
         }
   
         let result = []
         result.push(header, body, breaking, footer)
         result = result.filter(Boolean)
         const formatterResult = result.join('\n\n')
         return formatterResult
       })
     }
   }
   ```

4. 运行以下命令初始化 `commitizen` 规范。

   ```bash
   npx commitizen init cz-git --save-dev --save-exact
   ```

5. 配置 `husky` 钩子。

   先执行如下命令
   ```bash
   husky install   
   ```

   在项目路径`.husky`内创建 `commit-msg` 文件，并写入以下内容：

   ```shell
   #!/usr/bin/env sh
   . "$(dirname -- "$0")/_/husky.sh"
   
   npx --no -- commitlint --edit $1
   ```

6. 提交代码。

   现在，可以使用 `git cz` 命令来提交代码。

7. 配置 `cz-git` 以与 `open-ai` 一起使用。
    1. 安装 `czg`
       ```bash
       npm install -g czg
       ```
    2. https://platform.openai.com/account/api-keys
       登陆并创建你的 OpenAI API 密钥，通常以 sk- 开头
       运行命令 `npx czg --openai-token=<API secret key>` 填入 OpenAI API 密钥完成设置

       ```bash
       npx czg --openai-token=sk-xxxxx
       ```

    3. 运行 `npx czg ai`
        - 如需返回多个简短描述，并开启选择模式
       ```bash
       npx czg ai -N=5
       ```


CHANGELOG.md
#  (2023-02-25)


### Features

* danyow  2023/2/25  更新 `.czrc` 文件，以使用 `cz-git` 与 `commitizen` 一起工作 ([0aebce8](https://github.com/danyow/cz-test/commit/0aebce83fce00c4622de6282c132d0c20e671ac9))
* danyow  2023/2/25  配置 `husky` 钩子，`commitizen` 规范，以及 `commitlint` 规则，实现 git 提交规范化 ([39991a7](https://github.com/danyow/cz-test/commit/39991a79a5a4213ea7049259b92b879396d2b26c))





.commitlintrc.js
// .commitlintrc.js
const {execSync} = require("child_process");
const util = require("util");

// 获取用户名和当前日期
const author = execSync('git config user.name').toString().replace(/(\r\n\t|\n|\r\t)/g, '').trim()
const date = new Date().toLocaleDateString()

/** @type {import('cz-git').UserConfig} */

const issuePrefixes = [
  // 如果使用 tapd 作为开发管理
  {value: 'bug', name: 'bug:     标记缺陷已完成'},
  {value: 'story', name: 'story:   标记需求已完成'}
];

module.exports = {
  extends: ['@commitlint/config-conventional'],
  rules: {
    'subject-min-length': [2, 'always', 2],
    'subject-empty': [2, 'never'],
  },
  prompt: {
    alias: {fd: 'docs: fix typos'},
    messages: {
      type: '选择你要提交的类型 :',
      scope: '选择一个提交范围（可选）:',
      customScope: '请输入自定义的提交范围 :',
      subject: '填写简短精炼的变更描述 :\n',
      body: '填写更加详细的变更描述（可选）。使用 "|" 换行 :\n',
      breaking: '列举非兼容性重大的变更（可选）。使用 "|" 换行 :\n',
      footerPrefixesSelect: '选择关联tapd内容（可选）:',
      customFooterPrefix: '输入自定义tapd前缀 :',
      footer: '输入对应id (可选) :\n',
      confirmCommit: '是否提交或修改commit ?',
      generatingByAI: '正在生成你的 AI 提交信息...',
      generatedSelectByAI: '选择你认为合适的:',
    },
    types: [
      {value: 'feat', name: 'feat:     新增功能 | A new feature'},
      {value: 'fix', name: 'fix:      修复缺陷 | A bug fix'},
      {value: 'docs', name: 'docs:     文档更新 | Documentation only changes'},
      {value: 'style', name: 'style:    代码格式 | Changes that do not affect the meaning of the code'},
      {value: 'refactor', name: 'refactor: 代码重构 | A code change that neither fixes a bug nor adds a feature'},
      {value: 'perf', name: 'perf:     性能提升 | A code change that improves performance'},
      {value: 'test', name: 'test:     测试相关 | Adding missing tests or correcting existing tests'},
      {value: 'build', name: 'build:    构建相关 | Changes that affect the build system or external dependencies'},
      {value: 'ci', name: 'ci:       持续集成 | Changes to our CI configuration files and scripts'},
      {value: 'revert', name: 'revert:   回退代码 | Revert to a commit'},
      {value: 'chore', name: 'chore:    其他修改 | Other changes that do not modify src or test files'},
    ],
    useEmoji: false,
    emojiAlign: 'center',
    themeColorCode: '38;5;043',
    scopes: ['framework', 'lobby', 'music', 'health', 'work'],
    allowCustomScopes: true,
    allowEmptyScopes: true,
    customScopesAlign: 'bottom',
    customScopesAlias: 'custom',
    emptyScopesAlias: 'empty',
    upperCaseSubject: false,
    markBreakingChangeMode: false,
    allowBreakingChanges: ['feat', 'fix'],
    breaklineNumber: 100,
    breaklineChar: '|',
    skipQuestions: [],
    issuePrefixes: issuePrefixes,
    customIssuePrefixAlign: 'bottom',
    emptyIssuePrefixAlias: 'skip',
    customIssuePrefixAlias: 'custom',
    allowCustomIssuePrefix: false,
    allowEmptyIssuePrefix: true,
    confirmColorize: true,
    maxHeaderLength: Infinity,
    maxSubjectLength: Infinity,
    minSubjectLength: 0,
    scopeOverrides: undefined,
    defaultBody: '',
    defaultIssues: '',
    defaultScope: '',
    defaultSubject: '',
    aiNumber: 5,
    aiQuestionCB: ({
                     maxSubjectLength,
                     diff
                   }) => `用完整句子为以下 Git diff 代码写一个有见解并简洁的 Git 中文提交消息，不加任何前缀，并且内容不能超过 ${maxSubjectLength} 个字符: \`\`\`diff\n${diff}\n\`\`\``,
    formatMessageCB: (messageMod => {

      // 生成类型和范围
      let scope = (messageMod.scope === undefined || messageMod.scope.length === 0) ? '' : util.format('(%s)', messageMod.scope)

      let typeScope = util.format('%s%s:', messageMod.type, scope)

      // 生成header
      let subjects = []
      subjects.push(author, date, messageMod.subject)
      subjects = subjects.filter(Boolean)
      const subject = subjects.join('  ')

      // 生成header
      const header = util.format('%s %s', typeScope, subject)

      // 生成 body 和 breaking
      const body = messageMod.body
      const breaking = (messageMod.breaking === undefined || messageMod.breaking.length === 0) ? undefined : '重大更新：' + messageMod.breaking

      // 生成 footer
      let footer
      if (messageMod.footer !== undefined && messageMod.footer.length > 0) {
        try {
          const valuesArray = issuePrefixes.map(obj => obj.value);
          const values = valuesArray.join('|');

          const issuePattern = new RegExp(`(?:${values})`, 'g');
          const issue = messageMod.footer.match(issuePattern)

          // 尝试替换掉 颜色符号
          let idString = messageMod.footer.replaceAll(/\x1B\[\w*m/g, '')
          // 替换回车
          idString = idString.replaceAll('\n', '')

          const idPattern = /\b\d+\b/g;
          const ids = idString.match(idPattern);
          footer = util.format('--%s=%s --user=%s', issue[0], ids.join(','), author)
        } catch (ex) {
          console.log('生成 foot 失败：' + ex)

          footer = messageMod.footer.replaceAll('\n', '')
        }
      }

      let result = []
      result.push(header, body, breaking, footer)
      result = result.filter(Boolean)
      const formatterResult = result.join('\n\n')
      return formatterResult
    })
  }
}

